### PR TITLE
Files upgrade

### PR DESCRIPTION
**Pull request by rstoneback**
Heads up pull request that the files upgrade has begun. I've broken up the from_os class method into multiple methods that will be used to support parsing of remote filenames.

Pull started now to attach to the project.

### CODE REVIEW

**Comment by rstoneback** [comment]
Currently breaking in python3,
======================================================================
ERROR: Checking download routine functionality for module:  supermag magnetometer all 
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/test-environment/lib/python3.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/rstoneback/pysat/pysat/tests/test_instruments.py", line 236, in check_download
    raise e
  File "/home/travis/build/rstoneback/pysat/pysat/tests/test_instruments.py", line 213, in check_download
    password=user_download_dict[inst_name][1])
  File "/home/travis/build/rstoneback/pysat/pysat/_instrument.py", line 1290, in download
    password=password, **kwargs)
  File "/home/travis/build/rstoneback/pysat/pysat/instruments/supermag_magnetometer.py", line 710, in download
    raise ValueError(estr)
ValueError: unable to format station query for [2009]
-------------------- >> begin captured logging << --------------------
urllib3.connectionpool: DEBUG: Starting new HTTP connection (1): supermag.jhuapl.edu:80
urllib3.connectionpool: DEBUG: http://supermag.jhuapl.edu:80 "POST /mag/lib/services/??user=rstoneback&service=stations&fmt=ascii&year=2009 HTTP/1.1" 200 14479
--------------------- >> end captured logging << ---------------------

When I check on my local machine, in python3, things don't work either.
import pysat

mag = pysat.Instrument('supermag', 'magnetometer', tag='all', update_files=True)
pysat is searching for supermag magnetometer all files.
Found 3 of them.
When using this data please acknowledge the SuperMAG collaboration according to the request outlined in the metadata attribute 'acknowledgements'

etime = pysat.datetime(2009, 1, 1)
stime = pysat.datetime(2009, 1, 1)
mag.download(stime, etime, user='rstoneba@utdallas.edu')

Downloading data to:  /Users/russellstoneback/ScienceData/supermag/magnetometer/all/
Downloading SuperMAG data for 2009-01-01
When using this data please acknowledge the SuperMAG collaboration according to the request outlined in the metadata attribute 'acknowledgements'
Returning supermag magnetometer stations data for 01/01/09
---------------------------------------------------------------------------
UnicodeEncodeError                        Traceback (most recent call last)
<ipython-input-5-d5fac745ebe2> in <module>()
----> 1 mag.download(stime, etime, user='rstoneba@utdallas.edu')

/Users/russellstoneback/git/pysat/pysat/_instrument.py in download(self, start, stop, freq, user, password, **kwargs)
   1288                                data_path=self.files.data_path,
   1289                                user=user,
-> 1290                                password=password, **kwargs)
   1291         # get current file date range
   1292         first_date = self.files.start_date
/Users/russellstoneback/git/pysat/pysat/instruments/supermag_magnetometer.py in download(date_array, tag, sat_id, data_path, user, password, baseline, delta, options, file_fmt)
    761         # Save the file data
    762         with open(fname, "w") as local_file:
--> 763             local_file.write(out_data)
    764             local_file.close()
    765             del out_data
UnicodeEncodeError: 'ascii' codec can't encode characters in position 1319-1321: ordinal not in range(128) 

mag.files.files
Out[6]: 
2001-01-14    supermag_magnetometer_all_20010114.txt_2001-01-14
2001-01-15    supermag_magnetometer_all_20010115.txt_2001-01-15
2009-01-01    supermag_magnetometer_all_20090101.txt_2009-01-01
dtype: object


Checking stations, 
mag.download(stime, etime, user='rstoneba@utdallas.edu')
Downloading data to:  /Users/russellstoneback/ScienceData/supermag/magnetometer/stations/
Downloading SuperMAG stations for 2009
Updating pysat file list
pysat is searching for supermag magnetometer stations files.
Unable to find any files that match the supplied template. If you have the necessary files please check pysat settings and file locations (e.g. pysat.pysat_dir).
Updating instrument object bounds.

Something is wrong in the stations code, or rather, the files parsing area since that is where changes are. To be continued.

**Comment by rstoneback** [comment]
Now includes fixes for #50 and #70.  CDAWeb instrument support is there, time to copy things into existing instrument files.

**Comment by rstoneback** [comment]
I belive this pull has all the major features requested. I've broken up the from_os into more useful chunks, made them more accurate, and maintained files.from_os functionality. I added methods to the Instrument object to download all the data on the remote server but not on the local machine, as well as a command to return all of the remote files, with datetimes. I added a list_remote_files routine to the template instruments as well as the NASA CDAWeb methods routine. I expanded the remote files functionality to all CDAWeb instruments using that setup. Leaves Demeter, SuperDARN, SuperMAG, the sw indices, ICON, cosmic. Unit tests are coming.

**Comment by rstoneback** [comment]
and Madrigal....

Pull isn't quite ready but it is available to test out to see if it does what it needs to.

**Comment by rstoneback** [comment]
The pull request tests pass because the code is in good shape. The push tests aren't passing because of pyglow, once again. The rm -rf pyglow directory does not actually remove the directory, presumably because of travis, thus git won't clone pyglow onto the server... and we are stuck again. We'll have to happy with the pull tests alone on this one because I'm tired of messing with pyglow issues.

**Comment by rstoneback** [comment]
Added support to Madrigal and JRO and DMSP IVM files. The downside, just getting the remote file list for JRO will take about 2 hours. This is based on methods provided by madrigalWeb. I don't want to know how long it would take for DMSP.

**Comment by rstoneback** [comment]
I drifted into trying to solve the travis issue. Thought I had a good plan with a dual directory setup. It didn't work. So I just disabled the whole cache setup and put in a very straightforward clone and install. Still doesn't work. Not exactly sure what to do going forward. 

The positive part of this comment is I sorted out the documentation build bug. That works now. I added minimum versions to pandas and numpy since that is what numpy 0.11 said it needed. Fix for issue #119.

**Comment by rstoneback** [comment]
SuperMAG, SuperDARN, and Demeter remote file support has been added. The Super instruments have been faked and Demeter lets you know it doesn't work there. Leaves space weather indices, which should probably wait until a better organization is sorted out. Finally, we have omni hro.

How do we include a test for the download_updated_files method?... We don't have time to download all the data all the time...
I could put in a flag that goes through the code far enough to invoke the list_remote_files routine and process but not actually engage the download routine. The downside is the Madrigal instruments officially take forever. So while this works it doesn't work.

**Comment by jklenzing** [comment]
@rstoneback, we may need to re-examine the pyglow integration testing in travis.  There have been a number of updates to pyglow in the last four weeks:

https://github.com/timduly4/pyglow/pulls?q=is%3Apr+is%3Aclosed

Note that python 3 support is now added.

**Comment by aburrell** [comment]
@rstoneback is this pull request stable and ready for review?

**Comment by rstoneback** [comment]
Yes, tests pass except for the CDAWeb issue. Cod is ready for a review.

File: pysat/instruments/timed_see.py [hunk 1]
```diff
@@ -64,6 +64,9 @@
              'local_fname': fname}
 supported_tags = {'':{'': basic_tag}}
 download = functools.partial(cdw.download, supported_tags)
+# support listing files currently on CDAWeb
+list_remote_files = functools.partial(cdw.list_remote_files, 
+                               supported_tags=supported_tags)
```
Comment (aburrell) [inline pysat/instruments/timed_see.py]: Nonstandard alignment

File: pysat/_files.py [hunk 2]
```diff
@@ -147,9 +148,11 @@ def __init__(self, sat, manual_org=False, directory_format=None,
             self.data_path = self.data_path[:-1]
         elif self.data_path[-1] != os.path.sep:
             self.data_path = os.path.join(self.data_path, '')
-
+        
+        # store write to disk preference
         self.write_to_disk = write_to_disk
-        if write_to_disk is False:
+        if self.write_to_disk is False:
+            # using blank memory rather than loading from diisk
```
[the remark this answers is not in the document]
Comment (rstoneback) [reply pysat/_files.py]: Fixed

File: pysat/_files.py [hunk 3]
```diff
@@ -496,141 +491,404 @@ def from_os(cls, data_path=None, format_str=None,
...
+        'year', 'month', 'day', 'hour', 'min', 'sec', 'version', 'revision'
+        'files' - input list of files
+        
+    """
+ 
+    import collections
+
+    # create storage for data to be parsed from filenames
+    stored = collections.OrderedDict()
+    stored['year'] = []; stored['month'] = []; stored['day'] = [];
```
[the remark this answers is not in the document]
Comment (rstoneback) [reply pysat/_files.py]: Thank you. Fixed.

File: pysat/instruments/demeter_methods.py [hunk 4]
```diff
@@ -9,13 +9,37 @@
 import sys
 
 def download(date_array, tag, sat_id, data_path=None, user=None, password=None):
-    """ Download 
+    """ Download
```
[the remark this answers is not in the document]
Comment (rstoneback) [reply pysat/instruments/demeter_methods.py]: Thanks!

File: pysat/instruments/madrigal_methods.py [hunk 5]
```diff
@@ -232,6 +233,108 @@ def download(date_array, inst_code=None, kindat=None, data_path=None, user=None,
...
+    try:
+        format_str = supported_tags[sat_id][tag]
+    except KeyError:
+        raise ValueError('Problem parsing supported_tags')
+
+    # open connection to Madrigal
+    web_data = madrigalWeb.madrigalWeb.MadrigalData(url)
+    # get list of experiments for instrument from 1900 till now
+    now = pysat.datetime.now()
+    expList = web_data.getExperiments(inst_code,1900,1,1,0,0,0,now.year,
```
[the remark this answers is not in the document]
Comment (rstoneback) [reply pysat/instruments/madrigal_methods.py]: Fixed

File: pysat/instruments/jro_isr.py [hunk 6]
```diff
@@ -71,6 +71,11 @@
 madrigal_inst_code = 10
 madrigal_tag = {'':{'drifts':1910, 'drifts_ave':1911, 'oblique_stan':1800,
                     'oblique_rand':1801, 'oblique_long':1802},}
+
+# support listing files currently available on remote server (Madrigal)
+list_remote_files = functools.partial(mad_meth.list_remote_files, 
+                               supported_tags=supported_tags,
```
[the remark this answers is not in the document]
Comment (rstoneback) [reply pysat/instruments/jro_isr.py]: Fixed

File: pysat/instruments/iss_fpmu.py [hunk 7]
```diff
@@ -54,6 +54,9 @@
             'local_fname':fname}
 supported_tags = {'':{'':basic_tag}}
 download = functools.partial(cdw.download, supported_tags)
+# support listing files currently on CDAWeb
+list_remote_files = functools.partial(cdw.list_remote_files, 
+                               supported_tags=supported_tags)
```
[the remark this answers is not in the document]
Comment (rstoneback) [reply pysat/instruments/iss_fpmu.py]: Fixed

File: pysat/instruments/cnofs_ivm.py [hunk 8]
```diff
@@ -58,6 +59,9 @@
             'local_fname':ivm_fname}
 supported_tags = {'':{'':basic_tag}}
 download = functools.partial(cdw.download, supported_tags)
+# support listing files currently on CDAWeb
+list_remote_files = functools.partial(cdw.list_remote_files, 
+                               supported_tags=supported_tags)
```
[the remark this answers is not in the document]
Comment (rstoneback) [reply pysat/instruments/cnofs_ivm.py]: Fixed

File: pysat/instruments/cnofs_plp.py [hunk 9]
```diff
@@ -54,6 +54,9 @@
             'local_fname':fname}
 supported_tags = {'':{'':basic_tag}}
 download = functools.partial(cdw.download, supported_tags)
+# support listing files currently on CDAWeb
+list_remote_files = functools.partial(cdw.list_remote_files, 
+                               supported_tags=supported_tags)
```
[the remark this answers is not in the document]
Comment (rstoneback) [reply pysat/instruments/cnofs_plp.py]: Fixed

File: pysat/instruments/cnofs_vefi.py [hunk 10]
```diff
@@ -58,6 +58,9 @@
             'local_fname':fname}
 supported_tags = {'':{'dc_b':basic_tag}}
 download = functools.partial(cdw.download, supported_tags)
+# support listing files currently on CDAWeb
+list_remote_files = functools.partial(cdw.list_remote_files, 
+                               supported_tags=supported_tags)
```
[the remark this answers is not in the document]
Comment (rstoneback) [reply pysat/instruments/cnofs_vefi.py]: Fixed

File: pysat/instruments/dmsp_ivm.py [hunk 11]
```diff
@@ -82,7 +82,12 @@
                 'f16':{'':10116},
                 'f17':{'':10117},
                 'f18':{'':10118},}
-                
+
+# support listing files currently available on remote server (Madrigal)
+list_remote_files = functools.partial(mad_meth.list_remote_files, 
+                               supported_tags=supported_tags,
```
[the remark this answers is not in the document]
Comment (rstoneback) [reply pysat/instruments/dmsp_ivm.py]: Fixed

File: pysat/instruments/rocsat1_ivm.py [hunk 12]
```diff
@@ -58,7 +58,10 @@
             'local_fname':fname}
 supported_tags = {'':{'':basic_tag}}
 download = functools.partial(cdw.download, supported_tags)
-                
+# support listing files currently on CDAWeb
+list_remote_files = functools.partial(cdw.list_remote_files, 
+                               supported_tags=supported_tags)
```
[the remark this answers is not in the document]
Comment (rstoneback) [reply pysat/instruments/rocsat1_ivm.py]: Fixed

File: pysat/instruments/superdarn_grdex.py [hunk 13]
```diff
@@ -49,6 +50,83 @@
...
+        This object
+
+    Returns
+    --------
+    Void : (NoneType)
+        Object modified in place.
+    
+    
+    """
+
```
[the remark this answers is not in the document]
Comment (rstoneback) [reply pysat/instruments/superdarn_grdex.py]: Done

File: pysat/instruments/timed_see.py (hunk 1 above)
Comment (rstoneback) [reply pysat/instruments/timed_see.py]: Fixed

**Comment by rstoneback** [comment]
All review comments have been incorporated. I've resolved conflicts between the pull and develop. Tests are running.

**Comment by aburrell** [comment]
I verified the fails.  Merging.

**Pull request merged**